// @vxrn/vite-native-hmr/dist/esm/devServerClient.native.js
import prettyFormat from "pretty-format";
import { getDevServerLocation } from "./getDevServerLocation";
import { loadHMRClient } from "./hmr-client";
function _class_call_check(instance, Constructor) {
  if (!(instance instanceof Constructor))
    throw new TypeError("Cannot call a class as a function");
}
function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || !1, descriptor.configurable = !0, "value" in descriptor && (descriptor.writable = !0), Object.defineProperty(target, descriptor.key, descriptor);
  }
}
function _create_class(Constructor, protoProps, staticProps) {
  return protoProps && _defineProperties(Constructor.prototype, protoProps), staticProps && _defineProperties(Constructor, staticProps), Constructor;
}
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
loadHMRClient();
var DevServerClient = /* @__PURE__ */ function() {
  "use strict";
  function DevServerClient2() {
    var _this = this;
    _class_call_check(this, DevServerClient2), _define_property(this, "socket", void 0), _define_property(this, "buffer", []);
    var initSocket = function() {
      var address = `ws://${getDevServerLocation().host}/__client`;
      _this.socket = new WebSocket(address);
      var onClose = function(event) {
        console.warn("Disconnected from the Dev Server:", event.message), _this.socket = void 0;
      };
      _this.socket.onclose = onClose, _this.socket.onerror = onClose, _this.socket.onopen = function() {
        _this.flushBuffer();
      };
    };
    process.env.NODE_ENV === "development" && initSocket();
  }
  return _create_class(DevServerClient2, [
    {
      key: "send",
      value: function(level, data) {
        try {
          var _this_socket;
          (_this_socket = this.socket) === null || _this_socket === void 0 || _this_socket.send(JSON.stringify({
            type: "client-log",
            level,
            data: data.map(function(item) {
              return typeof item == "string" ? item : prettyFormat(item, {
                escapeString: !0,
                highlight: !0,
                maxDepth: 3,
                min: !0,
                plugins: [
                  // @ts-expect-error
                  prettyFormat.plugins.ReactElement
                ]
              });
            })
          }));
        } catch {
          try {
            var _this_socket1;
            (_this_socket1 = this.socket) === null || _this_socket1 === void 0 || _this_socket1.send(JSON.stringify({
              type: "client-log",
              level,
              data: data.map(function(item, index) {
                try {
                  return typeof item == "string" ? item : JSON.stringify(item);
                } catch (err) {
                  return `Error stringifying item at index ${index} - ${item} - ${err}`;
                }
              })
            }));
          } catch (err) {
            try {
              var _this_socket2;
              (_this_socket2 = this.socket) === null || _this_socket2 === void 0 || _this_socket2.send(JSON.stringify({
                type: "client-log",
                level: "error",
                data: [
                  "error sending client log: " + err
                ]
              }));
            } catch {
            }
          }
        }
      }
    },
    {
      // final err
      // Ignore error
      key: "flushBuffer",
      value: function() {
        var _this = this;
        globalThis.__vxrnTmpLogs && (globalThis.__vxrnTmpLogs.forEach(function(param) {
          var { level: level2, data: data2 } = param;
          _this.buffer.push({
            level: level2,
            data: data2
          });
        }), delete globalThis.__vxrnTmpLogs);
        var _iteratorNormalCompletion = !0, _didIteratorError = !1, _iteratorError = void 0;
        try {
          for (var _iterator = this.buffer[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = !0) {
            var { level, data } = _step.value;
            this.send(level, data);
          }
        } catch (err) {
          _didIteratorError = !0, _iteratorError = err;
        } finally {
          try {
            !_iteratorNormalCompletion && _iterator.return != null && _iterator.return();
          } finally {
            if (_didIteratorError)
              throw _iteratorError;
          }
        }
        this.buffer = [];
      }
    },
    {
      key: "log",
      value: function(level, data) {
        if (level !== "groupEnd")
          if (this.socket && this.socket.readyState === WebSocket.OPEN)
            this.flushBuffer(), this.send(level, data);
          else {
            if (globalThis.__vxrnTmpLogs) return;
            this.buffer.push({
              level,
              data
            });
          }
      }
    }
  ]), DevServerClient2;
}(), client = new DevServerClient(), setup = function() {
}, enable = function() {
}, disable = function() {
}, registerBundle = function() {
}, log = function(level, data) {
  client.log(level, data);
};
export {
  client,
  disable,
  enable,
  log,
  registerBundle,
  setup
};
